// bandersnatch/lib/esm/baseArg.js
export class BaseArg {
    constructor(name) {
        this.options = {};
        this.name = name;
    }
    /**
     * Set the argument/option description.
     */
    description(description) {
        this.options.description = description;
        return this;
    }
    /**
     * Whether this argument/option can be interactive.
     */
    isPromptable() {
        return !!this.options.prompt;
    }
    /**
     * Returns the prompt line.
     */
    getPrompt() {
        return typeof this.options.prompt === 'string'
            ? this.options.prompt
            : this.options.description
                ? this.options.description
                : this.name;
    }
    /**
     * Get default value, if specified.
     */
    getDefault() {
        return this.options.default;
    }
    /**
     * Get possible values, is specified.
     * @todo See if we can add this to autocompleter
     */
    getChoices() {
        return this.options.choices;
    }
    /**
     * Get type, is specified.
     */
    getType() {
        return this.options.type;
    }
    /**
     * Returns the argument/option identifier.
     */
    getName() {
        return this.name;
    }
    /**
     * Returns the argument/option description.
     */
    getDescription() {
        return this.options.description;
    }
    /**
     * Returns the argument/option options.
     */
    getOptions() {
        return this.options;
    }
}
